### ISSUE TITLE
Support Ease of Access

### ISSUE THREAD

**Issue by mick-nvaccess** [nvaccess/test_nvdaTracImport#21]
**Reported by jteh on 24 Apr 2009 01:58 UTC**
**Blocked by #5037**
**Blocking #100, #3854, #4158**
NVDA can currently only run on the user input desktop on which it was started. It cannot switch to secure input desktops, which is required for UAC screens and the Windows logon screen. For Windows Vista and later, the Ease of Access framework is used to launch ATs on other input desktops. NVDA should support Ease of Access.

**Comment by mick-nvaccess** [comment]
**Modified by jteh on 24 Apr 2009 01:59 UTC**

**Comment by mick-nvaccess** [comment]
**Comment by jteh on 18 May 2009 01:07 UTC**
The [service for Windows XP](ticket:97) works very well on Windows Vista and Windows 7. In addition, there are some issues blocking the implementation of useable support for EoA. We'll just use the service on Vista and Windows 7 unless there is a good reason to support EoA. Closing as wontfix for now; reopen later if the situation changes.

**Comment by mick-nvaccess** [comment]
**Modified by jteh on 18 May 2009 01:07 UTC**

**Comment by mick-nvaccess** [comment]
**Comment by jteh on 19 Jan 2014 15:05 UTC**
There are now a few good reasons to register with Ease of Access:
* Windows 8 and 8.1 intentionally delay all desktop apps set to start automatically by several seconds. This can be reduced somewhat with a registry tweak, but not entirely, and we shouldn't make changes that affect things other than NVDA anyway.
* ATs need to be registered with EoA in order to be candidates for activation with Windows+volumeUp.
* Aside from these, it's probably more future proof to do this and it means we don't have a background service running (no matter how light) even when NVDA isn't.

Windows 8 introduced changes to Ease of Access that make it quite usable for us. Specifically, you can now specify that an AT shouldn't be terminated on a desktop switch, which is exactly what we need.

The question is whether to do this for Windows Vista and 7 as well. Termination on desktop switch is annoying, but I've managed to come up with a convoluted way to work around that, though it does require another executable to be running while NVDA is running.

**Comment by mick-nvaccess** [comment]
**Modified by jteh on 19 Jan 2014 15:05 UTC**

**Comment by mick-nvaccess** [comment]
**Comment by James Teh <jamie@nvaccess.org> on 22 Jan 2014 06:10 UTC**
In [changeset:"fa75c1f08d967cfe59feebb3a6f2390d73148b5a"]:
```CommitTicketReference repository="" revision="fa75c1f08d967cfe59feebb3a6f2390d73148b5a"
Merge branch 't308' into next

Incubates #308.
```

**Comment by mick-nvaccess** [comment]
**Modified by jteh on 22 Jan 2014 06:16 UTC**

**Comment by mick-nvaccess** [comment]
**Comment by James Teh <jamie@nvaccess.org> on 29 Jan 2014 08:45 UTC**
In [changeset:"990146754c82183c0216f96c9361d49c5157f2d5"]:
```CommitTicketReference repository="" revision="990146754c82183c0216f96c9361d49c5157f2d5"
Merge branch 't308' into next

Incubates #308.
```

**Comment by mick-nvaccess** [comment]
**Comment by James Teh <jamie@nvaccess.org> on 30 Jan 2014 06:03 UTC**
In [changeset:"ab34b09d31615ba62a7b0d27e04e14dd3eed5a38"]:
```CommitTicketReference repository="" revision="ab34b09d31615ba62a7b0d27e04e14dd3eed5a38"
Merge branch 't308' into next

Incubates #308.
```

**Comment by mick-nvaccess** [comment]
**Comment by James Teh <jamie@nvaccess.org> on 30 Jan 2014 07:52 UTC**
In [changeset:"84a9a74400db875873fe11dd1dc09e03762481eb"]:
```CommitTicketReference repository="" revision="84a9a74400db875873fe11dd1dc09e03762481eb"
Merge branch 't308' into next

Incubates #308.
```

**Comment by mick-nvaccess** [comment]
**Comment by James Teh <jamie@nvaccess.org> on 31 Jan 2014 04:40 UTC**
In [changeset:"53050d1e9bd8c679e01bc8834354d793a393cedb"]:
```CommitTicketReference repository="" revision="53050d1e9bd8c679e01bc8834354d793a393cedb"
Merge branch 't308' into next

Incubates #308. I'll get this right eventually.
```

**Comment by mick-nvaccess** [comment]
**Comment by driemer.riemer@gmail.com on 1 Feb 2014 07:08 UTC**
I think this would be better to list under features, rather than bugs. 

 1. It is listed as an enhancement rather than a defect.
 2. It isn't something being fixed, rather something being added to nvda that speeds it up at start up.

I am not an expert on this though, so feel free to tell me if there are specific reasons it is listed as a bug fix.
Thanks,
Derek

**Comment by mick-nvaccess** [comment]
**Modified by jteh on 1 May 2014 00:23 UTC**

**Comment by mick-nvaccess** [comment]
**Comment by blindbhavya on 2 Oct 2014 10:04 UTC**
Hi.
I was wondering...
The status of this ticket is Incubating for months now.
IF the patch is good enough, I think the status should be made Fixed and Milestone 2014.4
Do correct me if I have missed anything.

**Comment by mick-nvaccess** [comment]
**Modified by blindbhavya on 2 Oct 2014 10:04 UTC**

**Comment by mick-nvaccess** [comment]
**Comment by jteh on 7 Oct 2014 23:11 UTC**
Replying to [driemer.riemer@…](comment:12):
> I think this would be better to list under features, rather than bugs. 
>  1. It is listed as an enhancement rather than a defect.
That's true as far as the code is concerned. However, from a user perspective, it fixes a bug which is that NVDA is slower to start automatically on Windows 8.1. Normally, the two audiences match, but in this case, they don't.

>  2. It isn't something being fixed, rather something being added to nvda that speeds it up at start up.
Again, think about the user perspective. The user doesn't think of it as a feature, even though it's a significant code refactor.

**Comment by mick-nvaccess** [comment]
**Comment by jteh on 7 Oct 2014 23:15 UTC**
Replying to [blindbhavya](comment:15):
> The status of this ticket is Incubating for months now.
> IF the patch is good enough, I think the status should be made Fixed and Milestone 2014.4
It's not good enough yet. It works, but the problem is that if you downgrade to an older version of NVDA, you'll end up with some serious issues such as NVDA not running in secure mode on secure desktops. This is because the installer for the old version doesn't know about Ease of Access. The solution is probably to make the installer use the older uninstaller, but that will only work for versions where that is implemented. For example, if we implement that for 2014.4, downgrading to 2014.3 will still break. There isn't really a good solution for this, but we should at least try to not break when downgrading to the previous version, which means making this change and then waiting one more release before we merge Ease of Access.

**Comment by mick-nvaccess** [comment]
**Comment by leonarddr on 3 Feb 2015 10:17 UTC**
Replying to [jteh](comment:18):
> There isn't really a good solution for this, but we should at least try to not break when downgrading to the previous version, which means making this change and then waiting one more release before we merge Ease of Access.

Is it really necessary to take care of people downgrading to older versions of NVDA? The only downgrading i've ever done is from a development branch to a stable branch. Also, older versions aren't easy to find on the NV Access website, so people would only downgrade to locally saved older versions or search in the sourceforge archives. People thus experienced wouldn't have any problem with applying some tweaks before downgrading. I'd say mentioning the downgrading issues in the release announcement would be enough.
If not, though, may be the installers recognision of ease of access could be implemented in 2015.2 so 2015.3 could be the milestone of this ticket?

**Comment by mick-nvaccess** [comment]
**Comment by jteh on 14 Apr 2015 07:02 UTC**
The plan is:
1. For downgrades, make the installer unregister the installation using the previous copy. This should be done for 2015.3 (it's too late for 2015.2). See #5037.
2. Because this could be a major security hazard, implement a new command line option (--eoa) which is only called when running from Ease of Access. This will cause older versions of NVDA to fail to start with an error. This is ugly, but it's much better than a security hazard.
3. Finally, get Ease of Access support into NVDA 2015.4. That at least means anyone downgrading to 2015.3 (the version before, which is most likely if any downgrading does occur) won't get bitten by the ugliness in (2).

**Comment by mick-nvaccess** [comment]
**Modified by jteh on 14 Apr 2015 07:03 UTC**

**Comment by mick-nvaccess** [comment]
**Comment by leonarddr on 14 Apr 2015 07:17 UTC**
Please note that, when #3724 will be ready to go in master, easeOfAccess can make use of winVersion.winVersion instead of retrieving it from sys

**Comment by mick-nvaccess** [comment]
**Comment by leonarddr on 18 Aug 2015 12:10 UTC**
Replying to [leonarddr](comment:23):
> Please note that, when #3724 will be ready to go in master, easeOfAccess can make use of winVersion.winVersion instead of retrieving it from sys

Please see the attached patch for this
